File: JDK11/opentraceability/src/main/java/opentraceability/models/events/kdes/EventKDECountry.java
package opentraceability.models.events.kdes;

import opentraceability.interfaces.IEventKDE;
import opentraceability.utility.Countries;
import opentraceability.utility.Country;
import opentraceability.utility.XElement;
import org.json.JSONObject;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilderFactory;


public class EventKDECountry extends IEventKDE {
    public Class valueType = Country.class;

    public Country value = null;

    public EventKDECountry() {
    }

    public EventKDECountry(String ns, String name) {
        this.namespace = ns;
        this.name = name;
    }

    @Override
    public JSONObject getJson() {
        throw new UnsupportedOperationException("Not implemented yet.");
    }

    @Override
    public XElement getXml() {
        Country value = this.value;
        if (value == null) return null;
        try {
            var element = new XElement(this.name);
            element.setValue(Integer.toString(value.iso));
            return element;
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return null;
    }

    @Override
    public void setFromJson(Object json) {
        String strValue = json.toString();
        this.value = Countries.parse(strValue);
    }

    @Override
    public void setFromXml(XElement xml) {
        String strValue = xml.getValue();
        this.value = Countries.parse(strValue);
    }

    @Override
    public String toString() {
        if (value != null) {
            return value.toString();
        } else {
            return "";
        }
    }
}

File: Kotlin/OpenTraceability/src/main/kotlin/opentraceability/models/events/kdes/eventkdecountry.kt
package opentraceability.models.events.kdes

import opentraceability.interfaces.IEventKDE
import opentraceability.utility.*
import org.json.*
import org.w3c.dom.Element
import java.lang.reflect.Type
import javax.xml.parsers.DocumentBuilderFactory
import kotlin.reflect.KType
import kotlin.reflect.full.createType

class EventKDECountry: EventKDEBase, IEventKDE {

    override var valueType: KType = Country::class.createType()

    var value: Country? = null

    constructor(){}

    constructor(ns: String, name: String) {
        this.namespace = ns;
        this.name = name;
    }


    override fun getJson(): JSONObject? {
        TODO("Not yet implemented")
    }

    override fun getXml(): Element? {
        val value = this.value ?: return null
        // you would typically use a XML parser here
        val document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument()
        val element = document.createElement(this.name)
        element.textContent = value.iso.toString()
        return element
    }

    override fun setFromJson(json: JSONObject) {
        val strValue = json.toString()
        this.value = Countries.parse(strValue)
    }

    override fun setFromXml(xml: Element) {
        this.value = Countries.parse(xml.textContent)
    }

    override fun toString(): String {
        if (value != null) {
            return value.toString()
        } else {
            return ""
        }
    }
}
